server/app/middlewares/websocket_backend.rb

```diff
@@ -67,29 +75,95 @@ def call(env)
     end
   end
 
-  ##
-  # On websocket connection open
-  #
-  # @param [Faye::WebSocket] ws
-  # @param [Rack::Request] req
-  def on_open(ws, req)
+  # @param node_id [String] request header Kontena-Node-Id
+  # @param grid_token [String] request header Kontena-Grid-Token
+  # @param init_attrs [Hash] initialize attributes on new node
+  # @raise [CloseError]
+  # @return [HostNode] with node_id set
+  def find_node_by_grid_token(node_id, grid_token, init_attrs)
     # check grid
-    grid = Grid.find_by(token: req.env['HTTP_KONTENA_GRID_TOKEN'].to_s)
-    unless grid
-      logger.error 'invalid grid token, closing connection'
-      ws.close(4001, "Invalid grid token")
-      return
+    grid = Grid.find_by(token: grid_token.to_s)
+
+    raise CloseError.new(4001), "Invalid grid token" unless grid
+
+    node = grid.host_nodes.find_by(node_id: node_id)
+
+    if !node
+      node = grid.host_nodes.create!(node_id: node_id, **init_attrs)
+
+      logger.info "new node #{node} connected using grid token"
+
+    elsif node.token
+      raise CloseError.new(4005), "Invalid grid token, node was created using a node token"
```

Comment:
Note that this doesn't prevent creating a node with `kontena node create`, and then configuring the agent with a grid token instead of the node token: the newly created node won't have any `node_id` yet, so the server will just create a new node for the agent. If the two nodes happen to have the same name, then the newly created node will get renamed by the server. The node created using `kontena node create` will just remain disconnected.

This only does anything if a node is provisioned using `kontena node create` + `KONTENA_NODE_TOKEN` and connects, and then later gets changed to use a grid token instead - that connection will get rejected.